data-prepper-plugins/cloudwatch-logs/src/main/java/org/opensearch/dataprepper/plugins/sink/client/CloudWatchLogsService.java

```diff
@@ -0,0 +1,221 @@
+        LOG.info("Attempting to push logs! {Batch size: " + buffer.getEventCount() + "}");
+        stopAndResetStopWatch();
+        startStopWatch();
+
+        ArrayList<InputLogEvent> logEventList = new ArrayList<>();
+        failedPost = true;
+
+        for (byte[] data: buffer.getBufferedData()) {
+            InputLogEvent tempLogEvent = InputLogEvent.builder()
+                    .message(new String(data))
+                    .timestamp(System.currentTimeMillis())
+                    .build();
+            logEventList.add(tempLogEvent);
+        }
+
+        while (failedPost && (failCounter < retryCount)) {
+            try {
+                PutLogEventsRequest putLogEventsRequest = PutLogEventsRequest.builder()
+                        .logEvents(logEventList)
+                        .logGroupName(logGroup)
+                        .logStreamName(logStream)
+                        .build();
+
+                cloudWatchLogsClient.putLogEvents(putLogEventsRequest);
+
+                requestSuccessCount.increment();
+                failedPost = false;
+
+                //TODO: When a log is rejected by the service, we cannot send it, can probably push to a DLQ here.
+
+            } catch (AwsServiceException | SdkClientException e) {
+                LOG.error("Failed to push logs with error: {}", e.getMessage());
+
+                Thread.sleep(calculateBackOffTime(backOffTimeBase));
+
+                LOG.warn("Trying to retransmit request... {Attempt: " + (++failCounter) + "}");
+                requestFailCount.increment();
+            }
+        }
+
+        buffer.clearBuffer();
+
+        if (failedPost) {
+            logEventFailCounter.increment(logEventList.size());
+            releaseEventHandles(false);
+            LOG.error("Error, timed out trying to push logs!");
+            throw new RetransmissionLimitException("Error, timed out trying to push logs! (Max retry_count reached: {" + retryCount + "})");
+        } else {
+            logEventSuccessCounter.increment(logEventList.size());
+            releaseEventHandles(true);
+            failCounter = 0;
+        }
+    }
+
+    private long calculateBackOffTime(long backOffTimeBase) {
+        return failCounter * backOffTimeBase;
+    }
+
+    private void runExitCheck() throws InterruptedException {
+        int bufferSizeWithOverHead = (buffer.getBufferSize() + (buffer.getEventCount() * LOG_EVENT_OVERHEAD_SIZE));
+        if ((thresholdCheck.isEqualToThresholdReached(bufferSizeWithOverHead, buffer.getEventCount()) && (buffer.getEventCount() > 0))) {
+            pushLogs();
+        }
+    }
+
+    private void releaseEventHandles(final boolean result) {
+        if (bufferedEventHandles.size() == 0) {
+            return;
+        }
+
+        for (EventHandle eventHandle : bufferedEventHandles) {
+            eventHandle.release(result);
+        }
+
+        bufferedEventHandles.clear();
+    }
+
+    private void startStopWatch() {
```

Comment:
You could create a small class which encapsulates the stop watch logic you are looking for. Basically, all of these private methods would be the public methods on that class. 

This snippet from above could also be encapsulated into a public method. Perhaps `startIfNotRunning()`.

```
            if (!stopWatchOn) {
                startStopWatch();
            }
```

Reply:
I have made a SinkStopWatch class that encapsulates this logic.